xor_cipher/extension.pyx

```diff
@@ -0,0 +1,109 @@
+from cpython.bytes cimport PyBytes_AsString, PyBytes_FromStringAndSize
+from cpython.bytearray cimport PyByteArray_AsString, PyByteArray_FromStringAndSize
+from cython import cdivision
+from libc.stdlib cimport malloc
+from libc.string cimport strcpy
+
+
+# NOTE: `cdivision(True)` removes python integer division checks
+
+
+cdef char* copy_bytes(bytes data):
+    cdef char* source = PyBytes_AsString(data)
+
+    cdef size_t length = len(data)
+
+    cdef char* destination = <char*> malloc((length + 1) * sizeof(char))  # `+ 1` for null-character
```

Comment:
Theoretically speaking, since we the size available at all times, wouldn't it be better to save an extra byte and use `memcpy` instead?

Reply:
Additionally, its reasonable to assume that the bytes the user wants to xor might contain a null byte in its contents, which would completely break this system.